packages/typespec-azure-resource-manager/src/resource.ts

```diff
@@ -652,7 +658,12 @@ export function getResourcePathElements(
       break;
     }
 
-    if (i + 1 < segments.length && isVariableSegment(segments[i + 1])) {
+    // if the next segment is the last segment
+    if (i + 1 === segments.length - 1 && isVariableSegment(segments[i + 1], singletonResourceName)) {
+      typeSegments.push(segments[i]);
+      instanceSegments.push(segments[i]);
+      instanceSegments.push(segments[i + 1]);
```

Comment:
so we only consider the singletonResourceName when the current segment is the last segement.